### PR TITLE
*: Add lifecycle events to events API

### LINKED ISSUE #4322: Extend events API to include lifecycle events
Right now the only way to track container lifecycle events is through parsing of the log messages on the events API. This is very unreliable and certainly not recommended.

Instead, we should introduce a new `lifecycle` message type and send messages for the following actions at least:

 - container-created
 - container-deleted
 - container-started
 - container-stopped
 - container-restarted
 - container-paused
 - container-updated
 - container-renamed
 - container-snapshot-created
 - container-snapshot-deleted
 - container-snapshot-renamed
 - container-snapshot-restored

The struct would probably be something like:

```go
type EventLifecycle struct {
    // Name of the action as listed above (e.g. container-renamed)
    Action string

    // URL to the affected object (e.g. /1.0/containers/foo)
    Source string

    // Relevant extra information (e.g. for rename, new_name: bar)
    Metadata map[string]interface{}    
}
```

To implement this, we'll be looking at changing:

 - shared/version.api.go (new API extension string)
 - doc/api-extensions.md (new API extension string)
 - doc/rest-api.md (example under /1.0/events)
 - lxd/api_1.0.go (new API extension string)
 - shared/api/event.go (new event type)
 - lxd/events.go (new helper function to send the new event)
 - lxd/container_lxc.go (source of most of the events above)
 - lxc/monitor.go (mention the new event type in the help)

### PR DESCRIPTION
**Pull request by monstermunchkin**
Resolves #4322

Signed-off-by: Thomas Hipp <thomas.hipp@canonical.com>

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/events.go [hunk 1]
```diff
@@ -44,6 +44,15 @@ func (h eventsHandler) Log(r *log.Record) error {
 	return nil
 }
 
+func SendLifecycleEvent(action, source string,
```
Comment (stgraber) [inline lxd/events.go]: SendLifecycleEvent should be renamed to eventSendLifecycle

File: shared/api/event.go [hunk 2]
```diff
@@ -18,3 +18,10 @@ type EventLogging struct {
 	Level   string            `yaml:"level" json:"level"`
 	Context map[string]string `yaml:"context" json:"context"`
 }
+
+// EventLifecycle represets a lifecycle type event entry
```
Comment (stgraber) [inline shared/api/event.go]: Should add an API extension comment here (grep for API in shared/api for examples).

File: doc/api-extensions.md [hunk 3]
```diff
@@ -443,3 +443,6 @@ The following existing endpoints have been modified:
  * `DELETE /1.0/storage-pool/<pool>/volumes/<type>/<name>` accepts a new target query parameter
  * `POST /1.0/networks` accepts a new target query parameter
  * `GET /1.0/networks/<name>` accepts a new target query parameter
+
+## lifecycle
```
Comment (stgraber) [inline doc/api-extensions.md]: Should be event_lifecycle to match api.go

**Comment by stgraber** [comment]
Some comments in-line and as discussed we should limit the context right now to be empty in all cases except for:

 - Rename should have the new name
 - All snapshot events should have the snapshot name

File: lxd/container_lxc.go [hunk 4]
```diff
@@ -3206,6 +3224,14 @@ func (c *containerLXC) Delete() error {
 
 	logger.Info("Deleted container", ctxMap)
 
+	if c.IsSnapshot() {
+		eventSendLifecycle("container-snapshot-deleted",
+			fmt.Sprintf("/1.0/containers/%s", c.name), nil)
```
Comment (stgraber) [inline lxd/container_lxc.go]: Missing the snapshot name

File: lxd/container_lxc.go [hunk 5]
```diff
@@ -3087,6 +3102,9 @@ func (c *containerLXC) Restore(sourceContainer container, stateful bool) error {
 		return nil
 	}
 
+	eventSendLifecycle("container-snapshot-restored",
+		fmt.Sprintf("/1.0/containers/%s", c.name), nil)
```
Comment (stgraber) [inline lxd/container_lxc.go]: Missing the snapshot name

File: lxd/container_lxc.go [hunk 6]
```diff
@@ -3326,6 +3352,18 @@ func (c *containerLXC) Rename(newName string) error {
 
 	logger.Info("Renamed container", ctxMap)
 
+	if c.IsSnapshot() {
+		eventSendLifecycle("container-snapshot-renamed",
+			fmt.Sprintf("/1.0/containers/%s", oldName), map[string]interface{}{
+				"new_name": newName,
```
Comment (stgraber) [inline lxd/container_lxc.go]: Missing the snapshot name

File: shared/api/event.go [hunk 7]
```diff
@@ -18,3 +18,12 @@ type EventLogging struct {
...
+//
+// API extension: event_lifecycle
+type EventLifecycle struct {
+	Action  string                 `yaml:"action" json:"action"`
+	Source  string                 `yaml:"source" json:"source"`
+	Context map[string]interface{} `yaml:"context" json:"context"`
```
Comment (stgraber) [inline shared/api/event.go]: Should be tagged "omitempty" I think, so that the rendering looks better when nil

**Comment by monstermunchkin** [comment]
Should be fine now.

**Pull request merged**